### PR TITLE
refactor(compiler-cli): support xi18n in ngtsc

### PR DESCRIPTION
**Pull request by alxhub**
xi18n is the operation of extracting i18n messages from templates in the
compilation. Previously, only View Engine was able to perform xi18n. This
commit implements xi18n in the Ivy compiler, and a copy of the View Engine
test for Ivy verifies that the results are identical.

### CODE REVIEW

**Comment by petebacondarwin** [comment]
@alxhub is the purpose of this PR to enable migration of G3 projects that use `ngc` to do the message extraction?

**Comment by alxhub** [comment]
> @alxhub is the purpose of this PR to enable migration of G3 projects that use `ngc` to do the message extraction?

Correct!

File: packages/compiler-cli/index.ts [hunk 1]
```diff
@@ -21,7 +21,9 @@ export * from './src/tooling';
 export {CompilerOptions as AngularCompilerOptions} from './src/transformers/api';
 
 export {ngToTsDiagnostic} from './src/transformers/util';
+
 export {NgTscPlugin} from './src/ngtsc/tsc_plugin';
 export {NgtscProgram} from './src/ngtsc/program';
 
+
```
Comment (petebacondarwin) [inline packages/compiler-cli/index.ts:28]: Unnecessary changes?

File: packages/compiler-cli/test/ngtsc/xi18n_spec.ts [hunk 2]
```diff
@@ -0,0 +1,337 @@
...
+const testFiles = loadStandardTestFiles({fakeCore: true, fakeCommon: true});
+
+runInEachFileSystem(os => {
+  let env!: NgtscTestEnvironment;
+
+  describe('ngtsc xi18n', () => {
+    if (os === 'Windows' || platform() === 'win32') {
+      // xi18n tests are skipped on Windows as the paths in the expected message files are platform-
+      // sensitive. These tests will be deleted when xi18n is removed, so it's not a major priority
+      // to make them work with Windows.
```
Comment (petebacondarwin) [inline packages/compiler-cli/test/ngtsc/xi18n_spec.ts:24]: What is the plan for deleting `xi18n` from the code base?

File: packages/compiler-cli/test/ngtsc/xi18n_spec.ts [hunk 3]
```diff
@@ -0,0 +1,337 @@
...
+
+      // Add a single disabled test to ensure the spec file is always non-empty.
+      xit('tests are disabled on Windows', () => {});
+      return;
```
Comment (petebacondarwin) [inline packages/compiler-cli/test/ngtsc/xi18n_spec.ts:28]: Could you just move the `if` block outside the `describe()` to avoid having to add this excluded test? E.g.

```ts
runInEachFileSystem(os => {
  if (os === 'Windows' || platform() === 'win32') return;

  describe('ngtsc xi18n', () => {
    ...
  });
});

```

**Review by petebacondarwin — APPROVED** [review]
LGTM! A couple of nits.

I am interested to hear what the plan is going forward for message extraction in G3.
Currently localization in G3 relies upon `goog.getMsg()` and closure compiler, right?
Outside G3, localization (and message extraction) is based on the `$localize` code.

File: packages/compiler-cli/test/ngtsc/xi18n_spec.ts (hunk 3 above)
Comment (alxhub) [reply packages/compiler-cli/test/ngtsc/xi18n_spec.ts:28]: Would this actually fix the problem? I thought the issue was an empty `_spec` file in general, not an empty `describe()` block.

Comment (petebacondarwin) [reply packages/compiler-cli/test/ngtsc/xi18n_spec.ts:28]: I just tried this locally by adding a `return;` statement directly after the `runInEachFileSystem(os => {` block of another file on master. It passed fine. But then I moved the `return;` inside the `describe()` block and it also passed. So I am not sure what the original problem was

**Review by atscott — APPROVED** [review]
reviewed-for: public-api

**Review by jessicajaniuk — APPROVED** [review]
LGTM :cookie:

Just a single error message suggestion that I don't feel strongly about.

reviewed-for: public-api

**Review by petebacondarwin — APPROVED** [review]
reviewed-for: public-api

File: packages/compiler-cli/test/ngtsc/xi18n_spec.ts (hunk 2 above)
Comment (alxhub) [reply packages/compiler-cli/test/ngtsc/xi18n_spec.ts:24]: `$localize` all the way!

File: packages/compiler-cli/src/transformers/i18n.ts [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+      return 'xmb';
+    case 'xlf':
+    case 'xlif':
+    case 'xliff':
+    case 'xlf2':
+    case 'xliff2':
+      return 'xlf';
+  }
+
+  throw new Error(`Unsupported format "${formatName}"`);
```
[the remark this answers is not in the document]
Comment (alxhub) [reply packages/compiler-cli/src/transformers/i18n.ts:29]: Agreed - given that this is a simple code move, though, I'd like to keep the original unchanged.

**Comment by alxhub** [comment]
Switching to `target: patch`, as this PR applies cleanly to the patch branch, and we want to keep the compiler in sync as much as possible.
